File: libtcalc/token.cpp
#include "token.h"

#include <array>
#include <format>

using namespace tc;

std::string_view tc::token_kind_name(token_kind type)
{
    using namespace std::literals;
    constexpr static std::array type_names =
    {
        "bad"sv,
        "end_of_file"sv,

        "numeric_literal"sv,
        "superscript_literal"sv,
        "hex_literal"sv,
        "binary_literal"sv,
        "identifier"sv,

        "plus"sv,
        "superscript_plus"sv,
        "minus"sv,
        "superscript_minus"sv,
        "multiply"sv,
        "divide"sv,
        "exponentiate"sv,
        "left_parenthesis"sv,
        "right_parenthesis"sv,
        "radical"sv,
        "percent"sv,
        "factorial"sv,
        "left_shift"sv,
        "right_shift"sv,
        "greater"sv,
        "greater_or_equal"sv,
        "less"sv,
        "less_or_equal"sv,
        "equal"sv,
        "equality"sv,
        "not_equal"sv,
        "pi"sv,
        "tau"sv,
        "binary_nand"sv,
        "binary_nor"sv,
        "binary_xnor"sv,
        "binary_and"sv,
        "binary_or"sv,
        "binary_xor"sv,
        "binary_not"sv,

        "argument_separator"sv,
        "end_of_line"sv,
    };

    return type_names.at(static_cast<size_t>(type));
}

std::string token::format() const
{
    std::string str = std::format("(L:{}, C:{}): {} ", position().line, position().column, token_kind_name(kind()));
    if (kind() != token_kind::end_of_line && kind() != token_kind::end_of_file)
        str += std::format("\"{}\"", source_str());
    return str;
}

File: tcalc/src/token.rs
use std::fmt;

use crate::source_pos::SourcePos;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Bad,
    EndOfFile,

    NumericLiteral,
    SuperscriptLiteral,
    HexLiteral,
    BinaryLiteral,
    Identifier,

    Plus,
    SuperscriptPlus,
    Minus,
    SuperscriptMinus,
    Multiply,
    Divide,
    Exponentiate,
    OpenParenthesis,
    CloseParenthesis,
    Radical,
    CubeRoot,
    FourthRoot,
    Percent,
    Factorial,
    LeftShift,
    RightShift,
    GreaterThan,
    GreaterOrEqual,
    LessThan,
    LessOrEqual,
    Equal,
    Equality,
    NotEqual,
    BinaryNand,
    BinaryNor,
    BinaryXnor,
    BinaryAnd,
    BinaryOr,
    BinaryXor,
    BinaryNot,

    Deg,
    Rad,
    Grad,

    ArgumentSeparator,
    ExpressionSeparator,
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{self:?}")
    }
}

#[derive(Debug, Clone)]
pub struct Token {
    pub source: String,
    pub position: SourcePos,
    pub kind: TokenKind,
}
